// @kikinteractive/kik/lib/scan-code.js
'use strict';

/**
 *  @class KikCode
 */
class KikCode {
}

/**
 *  @enum {number} KikCode.Colors
 **/
KikCode.Colors = {
    /**
     *  @constant
     *  @type {number}
     *  #42B4E6
     */
    KikBlue: 0,

    /**
     *  @constant
     *  @type {number}
     *  #42DFD8
     */
    Turquoise: 1,

    /**
     *  @constant
     *  @type {number}
     *  #24D7A7
     */
    Mint: 2,

    /**
     *  @constant
     *  @type {number}
     *  #25912B
     */
    Forest: 3,

    /**
     *  @constant
     *  @type {number}
     *  #87D300
     */
    KikGreen: 4,

    /**
     *  @constant
     *  @type {number}
     *  #F8CB1C
     */
    Sunshine: 5,

    /**
     *  @constant
     *  @type {number}
     *  #FC971B
     */
    OrangeCreamsicle: 6,

    /**
     *  @constant
     *  @type {number}
     *  #F9703A
     */
    BloodOrange: 7,

    /**
     *  @constant
     *  @type {number}
     *  #F7373C
     */
    CandyAppleRed: 8,

    /**
     *  @constant
     *  @type {number}
     *  #F88585
     */
    Salmon: 9,

    /**
     *  @constant
     *  @type {number}
     *  #F767C3
     */
    Coral: 10,

    /**
     *  @constant
     *  @type {number}
     *  #940D65
     */
    Cranberry: 11,

    /**
     *  @constant
     *  @type {number}
     *  #CB94FF
     */
    Lavender: 12,

    /**
     *  @constant
     *  @type {number}
     *  #8737F8
     */
    RoyalPurple: 13,

    /**
     *  @constant
     *  @type {number}
     *  #353CD4
     */
    Marine: 14,

    /**
     *  @constant
     *  @type {number}
     *  #5D7687
     */
    Steel: 15,

    /**
     *  @constant
     *  @type {number}
     *  #42B4E6
     */
};

module.exports = KikCode;
